docs/howtos/state_params_howto.ipynb

```diff
@@ -0,0 +1,289 @@
+        "        epsilon=1e-5,\n",
+        "        axis_name=\"batch\", # Name batch dim\n",
+        "    )\n",
+        "\n",
+        "    x = nn.Dense(self.hidden_size)(x)\n",
+        "    x = norm()(x)\n",
+        "    x = nn.relu(x)\n",
+        "    x = nn.Dense(self.hidden_size)(x)\n",
+        "    x = norm()(x)\n",
+        "    x = nn.relu(x)\n",
+        "    y = nn.Dense(self.out_size)(x)\n",
+        "\n",
+        "    return y"
+      ],
+      "metadata": {
+        "id": "r18oI6NOVp7Y"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Secondly, we need to specify the same name when calling `vmap` in our training code:"
+      ],
+      "metadata": {
+        "id": "H_3cI-2DWL3g"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def update_step(apply_fn, x_batch, y_batch, opt_state, params, state):\n",
+        "\n",
+        "  def batch_loss(params):\n",
+        "    def loss_fn(x, y):\n",
+        "      pred, updated_state = apply_fn(\n",
+        "        {'params': params, **state},\n",
+        "        x, mutable=list(state.keys())\n",
+        "      )\n",
+        "      return (pred - y) ** 2, updated_state\n",
+        "\n",
+        "    loss, updated_state = jax.vmap(\n",
+        "      loss_fn, out_axes=(0, None),  # Do not vmap `updated_state`.\n",
+        "      axis_name='batch'  # Name batch dim\n",
+        "    )(x_batch, y_batch)  # vmap only `x`, `y`, but not `state`.\n",
+        "    return jnp.mean(loss), updated_state\n",
+        "\n",
+        "  (loss, updated_state), grads = jax.value_and_grad(\n",
+        "    batch_loss, has_aux=True\n",
+        "  )(params)\n",
+        "\n",
+        "  updates, opt_state = tx.update(grads, opt_state)  # Defined below.\n",
+        "  params = optax.apply_updates(params, updates)\n",
+        "  return opt_state, params, updated_state, loss"
+      ],
+      "metadata": {
+        "id": "0jurc8PdWLVP"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Note that we also need to specify that the model state does not have a batch dimension. Now we are able to train the model:\n",
+        "\n"
+      ],
+      "metadata": {
+        "id": "4CI-cToTYURG"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "model = MLP(hidden_size=10, out_size=1)\n",
+        "variables = model.init(random.PRNGKey(0), dummy_input)\n",
+        "# Split state and params (which are updated by optimizer).\n",
+        "state, params = variables.pop('params')\n",
+        "del variables  # Delete variables to avoid wasting resources\n",
+        "tx = optax.sgd(learning_rate=0.02)\n",
+        "opt_state = tx.init(params)\n",
+        "\n",
+        "for _ in range(num_epochs):\n",
+        "  opt_state, params, state, loss = update_step(\n",
```

Comment:
It would be nice to print something. maybe print the loss after each epoch?

Reply:
Sure, I didn't add that earlier because I thought it would be inconsistent with the docs.